### PR TITLE
TPv2 bring Servers Detail/Table to parity with TPv1

### PR DESCRIPTION
**Pull request by shamrickus**
This PR bridges the gap of parity between TPv2 and v1 in terms of Servers Detail/Table doing a variety of things and in the process fixing #7485, #7484.

Specifically the following was changed/added:
* E2E tests
  * Extracting out the the E2E test data creation into it's own class and improving error messages so the user is aware of what request failed.
  * The test data creation is now run twice to ensure they are idempotent.
  * Upgrade the API version used by test data creation to ~~4.1~~ 4.0
  * The common table PO has been updated to add various methods that can be used to filter the table/select a row. The Servers Detail test have been updated to use these to navigate to a detail instead of using hardcoded urls.
  * Default window size of non-headless runs has been set.
* Updated to required node version to be what Angular 15 needs.
* Added CSS vars for toolbar/sidebar sizes.
* Servers Detail
  * Implemented update (previous was just a button that wasn't hooked up) & delete.
  * Implemented layered profiles
  * Implement New server.
  * Converted from UI design from a TP clone to something that looks similar but uses material components.
  * Added links to view details of fields such as CDN, Cache Group, etc.
  * Added help text in the form of suffix icon tooltips (MGMT fields)
  * Implement Queue/Clear Server Updates
* Generic Table
  * Fixed a bug  where the clicking the select label in the dropdown menus (such as column visibility) was not working and only worked if you clicked the select element, but not the label.
  * Added the ability to add "More" buttons. Moved `Resize` and `Export Grid` to the menu.
  * Added the ability to bind to the double click event on a grid row. 
  * Fixed a bug on dropdown menus where the arrow would not rotate when opening a menu (column visibility and now more)
* Servers Table
  * Add Create Server FAB
  * Added Queue/Dequeue buttons to the more menu (by cdn)
  * Added double click row to goto server details.

<hr/>

## Which Traffic Control components are affected by this PR?

- Documentation
- Traffic Portal v2

## What is the best way to verify this PR?
Ensure the changes work (listed above). Verify that the content works and looks good at various resolutions. Compare to TPv1 page to check parity.

## If this is a bugfix, which Traffic Control versions contained the bug?

- master

## PR submission checklist
- [x] This PR has tests 
- [ ] This PR has documentation 
- [ ] This PR has a CHANGELOG.md entry 
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://apache.org/security) for details)

### CODE REVIEW

File: experimental/traffic-portal/src/app/core/servers/servers-table/servers-table.component.ts [hunk 1]
```diff
@@ -320,23 +347,27 @@ export class ServersTableComponent implements OnInit {
...
 	 * Constructs the component with its required injections.
 	 *
+	 * @param auth The user authorization service.
 	 * @param api The Servers API which is used to provide row data.
 	 * @param route A reference to the route of this view which is used to set the fuzzy search box text from the 'search' query parameter.
-	 * @param router Angular router
 	 * @param navSvc Manages the header
+	 * @param cdn The CDN API
 	 * @param dialog Dialog manager
 	 */
```
Comment (ocket8888) [inline experimental/traffic-portal/src/app/core/servers/servers-table/servers-table.component.ts:361]: fwiw, you can just get rid of this. With the switch from TSLint to `typescript-eslint` it became possible to not require JSDoc comments on constructors.

File: experimental/traffic-portal/src/app/core/servers/update-status/update-status.component.html [hunk 2]
```diff
@@ -17,7 +17,7 @@ <h2 mat-dialog-title>Change status for {{serverName}}</h2>
 		<mat-form-field appearance="fill" *ngIf="true">
 			<mat-label>New Status</mat-label>
 			<select name="status" [(ngModel)]="status" required matNativeControl>
-				<option *ngFor="let status of statuses" [value]="status" [disabled]="status.id === currentStatus">{{status.name}}</option>
+				<option *ngFor="let status of statuses" [value]="status.name" [disabled]="status.id === currentStatus">{{status.name}}</option>
```
Comment (ocket8888) [inline experimental/traffic-portal/src/app/core/servers/update-status/update-status.component.html:20]: were these changes to the update-status component necessary? I don't necessarily disagree that they don't simplify it, since apparently the name is the only information we need, but it clutters the PR to do a refactor in the midst of such a large, unrelated change.

File: experimental/traffic-portal/src/app/api/cdn.service.ts [hunk 3]
```diff
@@ -78,6 +78,28 @@ export class CDNService extends APIService {
...
+		const action = {action: "queue"};
+		return this.post<CDNQueueResponse>(path, action).toPromise();
+	}
+
+	/**
+	 * Dequeue updates to servers by a CDN
+	 *
+	 * @param id The CDN id to dequeue server updates for
+	 */
+	public async dequeueServerUpdates(id: number): Promise<CDNQueueResponse> {
```
Comment (ocket8888) [inline experimental/traffic-portal/src/app/api/cdn.service.ts:97]: these should also be able to take in `ResponseCDN` objects in addition to just the IDs thereof

**Review by ocket8888 — CHANGES_REQUESTED** [review]

File: experimental/traffic-portal/nightwatch/dataClient.ts [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+			typeId: httpType.id,
+			xmlId: `testDS${id}`
+		};
+		url = `${apiUrl}/deliveryservices`;
+		try {
+			resp = await this.client.post(url, JSON.stringify(ds));
+		} catch(e) {
+			(e as AxiosError).message += ` Failed on DS post at ${url}`;
+			throw e;
+		}
```
[the remark this answers is not in the document]
Comment (shamrickus) [reply experimental/traffic-portal/nightwatch/dataClient.ts:186]: You are correct, I had it doing more stuff but all of that info is already on the error.

File: experimental/traffic-portal/src/app/core/misc/isogeneration-form/isogeneration-form.component.html [hunk 5]
```diff
@@ -79,31 +79,6 @@
...
-						<input matInput type="text" formControlName="mgmtIpGateway"/>
-						<mat-error>Invalid IPv4 address</mat-error>
-					</mat-form-field>
-					<mat-form-field>
-						<mat-label>Management IPv4 Netmask</mat-label>
-						<input matInput type="text" formControlName="mgmtIpNetmask"/>
-						<mat-error>Invalid IPv4 Netmask</mat-error>
-					</mat-form-field>
-				</fieldset>
-			</mat-card>
```
[the remark this answers is not in the document]
Comment (shamrickus) [reply experimental/traffic-portal/src/app/core/misc/isogeneration-form/isogeneration-form.component.html:106]: It appears you deployed the types incorrectly. I see the fields on the repo, but the v4.0.9 release _definitely_ does not have the mgmt fields.

File: experimental/traffic-portal/src/app/core/servers/server-details/server-details.component.spec.ts [hunk 6]
```diff
@@ -157,16 +157,20 @@ describe("ServerDetailsComponent", () => {
...
 		expect(() => component.changeStatus(new MouseEvent("click"))).toThrow();
+*/
 	});
 
 	it("closes the 'change status' dialog when done", () => {
+		/*
 		component.changeStatusDialogOpen = true;
 		component.doneUpdatingStatus(true);
 		expect(component.changeStatusDialogOpen).toBeFalse();
+*/
```
[the remark this answers is not in the document]
Comment (shamrickus) [reply experimental/traffic-portal/src/app/core/servers/server-details/server-details.component.spec.ts:174]: Meant to fix these before pushing.

File: CHANGELOG.md [hunk 7]
```diff
@@ -53,7 +53,6 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - [#7469](https://github.com/apache/trafficcontrol/pull/7469) *Traffic Ops* Changed logic to not report empty or missing cookies into TO error.log.
 
 ### Fixed
-- [#7505](https://github.com/apache/trafficcontrol/pull/7505) *Traffic Portal* Fix an issue where a Delivery Service with Geo Limit Countries Set was unable to be updated.
```
Comment (ocket8888) [inline CHANGELOG.md:56]: bad merge

**Review by ocket8888 — CHANGES_REQUESTED** [review]

File: experimental/traffic-portal/src/app/api/cdn.service.ts [hunk 8]
```diff
@@ -78,6 +78,30 @@ export class CDNService extends APIService {
...
+	 *
+	 * @param cdn The CDN or ID to dequeue from
+	 */
+	public async dequeueServerUpdates(cdn: ResponseCDN | number): Promise<CDNQueueResponse> {
+		const id = typeof cdn === "number" ? cdn : cdn.id;
+		const path = `cdns/${id}/queue_update`;
+		const action = {action: "dequeue"};
+		return this.post<CDNQueueResponse>(path, action).toPromise();
+	}
+
```
Comment (ocket8888) [inline experimental/traffic-portal/src/app/api/cdn.service.ts:104]: These need tests

File: experimental/traffic-portal/src/app/api/testing/server.service.ts [hunk 9]
```diff
@@ -41,7 +41,7 @@ function serverCheck(server: ResponseServer): Servercheck {
 		cacheGroup: server.cachegroup ?? "SERVER HAD NO CACHE GROUP",
 		hostName: server.hostName ?? "SERVER HAD NO HOST NAME",
 		id: server.id,
-		profile: server.profile ?? "SERVER HAD NO PROFILE",
+		profile: server.profileNames.length > 0 ? server.profileNames[0] :  "SERVER HAD NO PROFILE",
```
Comment (ocket8888) [inline experimental/traffic-portal/src/app/api/testing/server.service.ts:44]: We don't have `noUncheckedIndexedAccess` enabled, so fwiw you can just do `server.profileNames[0] ?? "SERVER HAD NO PROFILE"`

File: experimental/traffic-portal/src/styles.scss [hunk 10]
```diff
@@ -100,7 +102,122 @@ body {
...
+			width: 16px !important;
+			height: 16px !important;
+		}
+	}
+
+	.mat-mdc-button-touch-target {
+		width: 24px !important;
+		height: 24px !important;
+	}
+}
```
Comment (ocket8888) [inline experimental/traffic-portal/src/styles.scss:134]: gosh that's an awful lot of `!important` - is that really necessary? And if so: why?

**Review by ocket8888 — CHANGES_REQUESTED** [review]

File: experimental/traffic-portal/src/styles.scss (hunk 10 above)
Comment (shamrickus) [reply experimental/traffic-portal/src/styles.scss]: Because of how the material CSS is encaspulated. If it's put in the component scss it would work but this is going to be needed else where. I added a mixin that should alleviate the issue

File: experimental/traffic-portal/src/app/api/testing/server.service.ts [hunk 11]
```diff
@@ -460,4 +487,21 @@ export class ServerService {
...
 		return created;
 	}
+
+	/**
+	 * Deletes an existing server.
+	 *
+	 * @param server The Server to be deleted, or just its ID.
+	 * @returns The deleted server.
+	 */
+	public async deleteServer(server: number | ResponseServer): Promise<ResponseServer> {
```
Comment (ocket8888) [inline experimental/traffic-portal/src/app/api/testing/server.service.ts:497]: the methods added to this service need test coverage

**Review by ocket8888 — CHANGES_REQUESTED** [review]

**Review by ocket8888 — APPROVED** [review]
this looks good, just needs to be rebased

File: experimental/traffic-portal/build/package-lock.json [hunk 12]
```diff
@@ -12,7 +12,7 @@
         "pm2": "^5.2.2"
       },
       "engines": {
-        "node": ">=16.20"
+        "node": ">=16.14"
```
Comment (ocket8888) [inline experimental/traffic-portal/build/package-lock.json:15]: was lowering the version requirement intentional?

Comment (shamrickus) [reply experimental/traffic-portal/build/package-lock.json:15]: Yes, `16.14` is the real required version
